lib/preload.js

```diff
@@ -382,13 +382,56 @@ class Preloader extends EventEmitter {
 			.catch((error) => this.emit('error', error));
 	}
 
+	/**
+	 * Send a command to `preload.py` (running in a Docker container) through
+	 * its stdin, and read a response from its stdout. The stdout response
+	 * must be a JSON object containing a `statusCode` field (integer) and
+	 * optionally `result` and `error` fields. If `statusCode` is the number
+	 * zero, the command execution is considered successful, otherwise it is
+	 * assumed to have failed and the returned promise is rejected with an
+	 * error message including the message provided in the `error` field.
+	 *
+	 * @param {string} command
+	 * @param {object} parameters
+	 */
 	_runCommand(command, parameters) {
 		return new Bluebird((resolve, reject) => {
+			const cmd = JSON.stringify({ command, parameters }) + '\n';
 			this.stdout.once('error', reject);
```

Comment:
Can you confirm my thinking that dockerode/Bluebird communication errors (not related to our python script) are caught by this `('error', reject)` function? It looks good to me and that's why we don't want to mix up our in-band `statusCode` with the out-of-band `StatusCode`.

Reply:
My understanding is limited -- including my understanding of your question, 😁   not sure exactly what to make of "dockerode/Bluebird communication errors". I am tempted to say that the answer is 'no'. Focusing specifically this line:
```js
	this.stdout.once('error', reject);
```
Through this line, `reject` will be called on errors emitted by the `this.stdout` stream (obviously). And in fact only the very first error (that gets emitted after the `reject` handler is attached), because the line uses `.once()` instead of `.on()`. Going back to where `this.stdout` is attached:
https://github.com/balena-io-modules/balena-preload/blob/v10.4.4/lib/preload.js#L879
```js
	this._prepareErrorHandler();
	return this.container.attach({ stream: true, stdout: true, stderr: true, stdin: true });
}).then((stream) => {
	this.stdin = stream;
	this.docker.modem.demuxStream(stream, this.stdout, this.stderr);
```
So I gather that `this.stdout` is the stdout of the container where the Python script is running. However, looking at the source source of `docker-modem` (v1.0.9 as actually currently used by `balena preload`), the implementation of `demuxStream`:
https://github.com/apocas/docker-modem/blob/v1.0.9/lib/modem.js#L304
```
Modem.prototype.demuxStream = function(stream, stdout, stderr) {
	...
	stdout.write(payload);
```
... we see there is no `.pipe(stdout)` connecting a "source of error events" to `this.stdout`, only the occasion `stdout.write(payload)` where `payload` is probably a `Buffer` object.

So I think that `this.stdout.once('error', reject);` might never actually produce any errors, `reject` might never actually be called on that line.

But so what? :-)  Why does it matter?

Note that the are other rejection routes at play. The high level `prepare()` and `preload()` functions in `preload.js` return promises that may reject at various points, and also the `Preloader` class itself is a subclass of `EventEmitter`:

```js
    class Preloader extends EventEmitter {
```

At the CLI level, there is the following line:

https://github.com/balena-io/balena-cli/blob/v12.44.12/lib/commands/preload.ts#L285
```js
    preloader.on('error', reject);
```

An in PR #241, I added a missing `this.emit()`:

https://github.com/balena-io-modules/balena-preload/pull/241/files#diff-337c0a7cc97920afbd5531862b39c73752b1377bbda24c26add1e511c63a2cc6R382
```
   .catch((error) => this.emit('error', error));
```

In `preload.js`, there are now 2 occurrences of `this.emit('error', error);`, and if any is called, the CLI should catch it at the top level.

I am not stating that we have covered all possible sources of errors. When you mentioned "dockerode/Bluebird communication errors", what specific `dockerode` call did you have in mind?

Reply:
Oh I see, Bluebird is just a promise library. I thought it was part of the dockerode communication module. In that case it makes sense that `this.stdout.once('error', reject);` may never get called. I even tried returning data on stderr in the python script and it still triggered the `data` handler and not the `error` handler.